File: 347-Top-K-Frequent-Elements/solution.ts
/**
 * Hash Map & Bucket Array
 * Time is O(n) | Space O(n)
 * https://leetcode.com/problems/top-k-frequent-elements
 *
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
function topKFrequent(nums: number[], k: number): number[] {
  const freqMap: Map<number, number> = new Map();
  const bucket: Set<number>[] = [];
  const result: number[] = [];

  for (let num of nums) {
    freqMap.set(num, (freqMap.get(num) || 0) + 1);
  }

  for (let [num, freq] of freqMap) {
    bucket[freq] = (bucket[freq] || new Set()).add(num);
  }

  for (let i = bucket.length - 1; i >= 0; i--) {
    if (bucket[i]) result.push(...bucket[i]);
    if (result.length === k) break;
  }
  return result;
}

File: 347-Top-K-Frequent-Elements/solution.go
package solutions

func topKFrequent(nums []int, k int) []int {
	freq := make(map[int]int, len(nums))

	for _, num := range nums {
		freq[num]++
	}

	count := make([][]int, len(nums)+1)

	for num, freq := range freq {
		count[freq] = append(count[freq], num)
	}

	res := make([]int, 0, k)

	for i := len(count) - 1; i >= 0; i-- {
		for _, num := range count[i] {
			res = append(res, num)
			if len(res) == k {
				return res
			}
		}
	}

	return res
}
